### ISSUE TITLE
Pagination caching

### ISSUE THREAD

**Issue by rarkins** [renovatebot/renovate#3837]
Following #3163, I'd like to add intelligent caching of paginated requests.

Background: there are repos that might have 1000s of closed Renovate PRs, but we need to know all of these if we are to implement our "ignore" feature. The problem is we have to retrieve them one-by-one using pagination and it could even take a minute to do so in some cases.

Instead, I have this idea:
1. We persistent cache our results, e.g. the "closed PR list" in this example, using just the fields we care about.
2. We make sure to include the lastModifiedDate as one of these fields
3. Based on our previous run, we know what the most recent last-modified-date was in the PR list, e.g. could be 8 minutes prior to Renovate's last run.
4. We pass that last modified date to got's pagination, with an option that says "paginate until this last modified date"
5. The got pagination will retrieve the first page of results, check if the oldest last modified date is before the "paginate until" date, and stop if it is.

I think in 99% of cases we won't see more than 100 (i.e. one page) changes between runs, which should mean that we don't even need to paginate - the first page should be fine.

Obviously this feature depends specifically on GitHub's last modified date being accurate, although the feature could later be made more generic in got if there was a field or combination of fields that you could rely on (e.g. could be an ID if they are incrementing and write-only, etc).

@szmarczak I thought you might be interested in this idea too. It's kind of a nice use cache but would be suuuuuuuuper useful when used.

**Comment by rarkins** [comment]
@psyb0t we should also implement this for GraphQL too, once we start retrieving PRs via GraphQL recursion. I think for GraphQL that the server doesn't return pagination instructions in the Link header, right? Instead we manage it more at the app level by looking at the response and cursor

**Comment by psyb0t** [comment]
@rarkins yep, it's more at the app level.

**Comment by szmarczak** [comment]
@rarkins Indeed, this would be very useful. While the idea itself looks quite simple, the API would be quite complicated. It needs to filter elements, needs to know if it should continue (when to stop, where to start from) etc.

**Comment by rarkins** [comment]
Blocked by #11419

**Comment by HonkingGoose** [comment]
The blocking issue is closed now:

- #11419

Are we ready to start working on pagination caching now?

**Comment by viceice** [comment]
i think this is already done by @zharinov.

@zharinov do you agree?

**Comment by zharinov** [comment]
Yes, I think we've done it
